drivers: sensor: temperature: Add i.MX RT die temperature sensor

anedergaard opened the pull request:
- Added dts binding for die temperature sensor fori.MX RT118X 
- Added driver for die temperature readings on i.MX RT118X

anedergaard committed:
dts: bindings: Add nxp i.MX RT die temperature binding

Added die temperature binding for i.MX RT 118X

Signed-off-by: Anders Bjørn Nedergaard <abn@polytech.com>

anedergaard committed:
drivers: sensor: temperature: Add i.MX RT die temperature sensor

Added driver for i.MX RT118X die temperature sensor

Signed-off-by: Anders Bjørn Nedergaard <abn@polytech.com>

decsny commented on dts/bindings/sensor/nxp,imxrt-tmpsns.yaml:
```diff
@@ -0,0 +1,8 @@
+
+description: NXP TMPSNS
+
+compatible: "nxp,imxrt-temperature"
```
please no soc family name in the compatible , @DerekSnell can you please suggest compatible string

jeppenodgaard commented on drivers/sensor/nxp/nxp_imxrt_temp/temp_imxrt.c:
```diff
@@ -0,0 +1,68 @@
+	float die_temp;
+};
+
+static int tmpsnsChannelGet(const struct device *dev, enum sensor_channel chan,
```
Replace all camelCase with snake_case

jeppenodgaard commented on drivers/sensor/nxp/nxp_imxrt_temp/temp_imxrt.c:
```diff
@@ -0,0 +1,68 @@
+
+	const struct imxrt_tmpsns_data *data = dev->data;
+
+	val->val1 = (int32_t)data->die_temp;
```
Use `sensor_value_from_float`

jeppenodgaard commented on drivers/sensor/nxp/nxp_imxrt_temp/temp_imxrt.c:
```diff
@@ -0,0 +1,68 @@
+
+static int tmpsnsChannelFetch(const struct device *dev, enum sensor_channel chan)
+{
+	if (chan != SENSOR_CHAN_DIE_TEMP) {
```
&& SENSOR_CHAN_ALL

jeppenodgaard reviewed (changes requested):
Functionality LGTM.
I think some naming alignment is needed. There's tmpsns, temp, temperature.

anedergaard commented:
> Functionality LGTM. I think some naming alignment is needed. There's tmpsns, temp, temperature.

Yes, you are right, I'll align the naming once @DerekSnell has suggested a compatible string.

DerekSnell commented:
Hi @anedergaard ,
Thank you for this contribution, and for getting guidance on the naming.  As @decsny stated, NXP does not like to use SOC or SOC family names in the bindings and drivers, because the hardware IP is frequently re-used across the portfolio on other familes.

The IP name for this HW peripheral is TMPSNS, which is also the name used in the SOC reference manual, and in the HAL driver [fsl_tempsensor.c](https://github.com/zephyrproject-rtos/hal_nxp/blob/master/mcux/mcux-sdk/drivers/tempsensor/fsl_tempsensor.c).  So I think the best name for this binding is `nxp,tmpsns`.

Thank you

anedergaard commented:
Hi @DerekSnell,
Thank you for the feedback on naming and the reasoning behind it - I appreciate it!

Updated 1
- Addressed driver naming, renamed binding to `nxp,tmpsns`
- Rebased to address conflicts

jeppenodgaard commented on drivers/sensor/nxp/nxp_tmpsns/nxp_tmpsns.c:
```diff
@@ -0,0 +1,68 @@
+	const struct nxp_tmpsns_data *data = dev->data;
+
+	switch (chan) {
+	case SENSOR_CHAN_ALL:
```
Delete this line.
`get` cannot support `SENSOR_CHAN_ALL` since there is only one `val` reference.

jeppenodgaard reviewed (changes requested):
Naming looks good :)

DerekSnell reviewed (changes requested):
Hi @anedergaard ,

Thank you for contributing this.  I am blocking to address this build warning when I build the test for the MIMXRT1170-EVK.  The test does build and runs well on my board.  But can you improve the DTS to avoid this warning?

```
CMake Warning at 
/zephyrproject/zephyr/cmake/modules/dts.cmake:425 (message):
  dtc raised one or more warnings:
 
  /build/zephyr/zephyr.dts:4051.27-4054.5:
  Warning (simple_bus_reg): /soc/temp: missing or empty reg/ranges property
```

Otherwise, the naming changes you made look good to me.

I see your PR got caught in a known CI issue with the RT1180, which should be resolved when https://github.com/zephyrproject-rtos/zephyr/pull/83827 merges.

A very minor nit, but since you will be updating - there is a typo in commit message "[termerature](https://github.com/zephyrproject-rtos/zephyr/pull/83880/commits/14b52e0c897fb64f4fcb251b3ad05a808bb65fb9)"

Thank you

MaureenHelm commented on dts/bindings/sensor/nxp,tmpsns.yaml:
```diff
@@ -0,0 +1,8 @@
+
+compatible: "nxp,tmpsns"
+
+include: base.yaml
```
```suggestion
include: sensor-device.yaml
```

anedergaard replied:
Thanks @MaureenHelm, it should now be placed in the correct commit.

MaureenHelm commented on drivers/sensor/nxp/nxp_tmpsns/CMakeLists.txt:
```diff
@@ -0,0 +1,3 @@
+zephyr_library()
```
File needs a license/copyright header

mmahadevan108 commented on drivers/sensor/nxp/nxp_tmpsns/Kconfig:
```diff
@@ -0,0 +1,13 @@
+	bool "NXP TMPSNS sensor"
+	depends on SENSOR
+	depends on DT_HAS_NXP_TMPSNS_ENABLED
+	depends on SOC_SERIES_IMXRT11XX || SOC_SERIES_IMXRT118X
```
Do we need this `depends`. Is'nt is sufficient to base it on if the temperature sensor is enabled in device tree.

anedergaard replied:
Removed the dependency to SOC_SERIES_IMXRT

mmahadevan108 commented on drivers/sensor/nxp/nxp_tmpsns/nxp_tmpsns.c:
```diff
@@ -0,0 +1,67 @@
+	switch (chan) {
+	case SENSOR_CHAN_ALL:
+	case SENSOR_CHAN_DIE_TEMP:
+		data->die_temp = TMPSNS_GetCurrentTemperature(TMPSNS);
```
You should be getting the base address from device tree.

mmahadevan108 commented on dts/arm/nxp/nxp_rt118x.dtsi:
```diff
@@ -44,6 +44,13 @@
+
+	soc {
+		temp: temp {
+			compatible = "nxp,tmpsns";
```
Need to add base address and interrupt properties. Also this should be part of the peripheral block

mmahadevan108 commented on dts/arm/nxp/nxp_rt11xx.dtsi:
```diff
@@ -1307,6 +1307,11 @@
 		};
+
+		temp: temp {
+			compatible = "nxp,tmpsns";
```
Missing base and interrupt properties

yperess commented on drivers/sensor/nxp/nxp_tmpsns/nxp_tmpsns.c:
```diff
@@ -0,0 +1,67 @@
+#include <zephyr/drivers/sensor.h>
+
+struct nxp_tmpsns_data {
+	float die_temp;
```
If you're going to use a `float` instead of the `q31` format you should depend on `FPU`

anedergaard replied:
Will add dependency on FPU

yperess commented on drivers/sensor/nxp/nxp_tmpsns/nxp_tmpsns.c:
```diff
@@ -0,0 +1,67 @@
+	config.measureMode = kTEMPSENSOR_ContinuousMode;
+	config.frequency = 0x03U;
+	TMPSNS_Init(TMPSNS, &config);
+	TMPSNS_StartMeasure(TMPSNS);
```
Is it necessary to start measuring on init? Why not enable the attribute for sample frequency to start sampling?

anedergaard replied:
Was done in init to allow parsing the `poll` part of the test in `tests/drivers/sensor/temp_sensor`

anedergaard commented:
Thanks for the review feedback, I appreciate it.
I'll look into the requested changes.

decsny commented:
@anedergaard any update for this PR?

anedergaard commented:
> @anedergaard any update for this PR?

Hi @decsny, sorry for stalling this PR.

I will address the hal_nxp SDK-NG changes first, then address the changes requested for this PR

anedergaard commented:
Update: Moved to use hal_nxp SDK-NG

anedergaard commented:
Rebased to rerun CI checks after CI issue

anedergaard commented:
Rebased to trigger CI workflow

MaureenHelm commented on drivers/sensor/nxp/nxp_tmpsns/Kconfig:
```diff
@@ -0,0 +1,13 @@
+
+config NXP_TMPSNS
+	bool "NXP TMPSNS sensor"
+	depends on SENSOR
```
This is redundant and can be removed

anedergaard replied:
removed redundant `depends on SENSOR`

MaureenHelm commented on tests/drivers/sensor/temp_sensor/boards/mimxrt1180_evk_mimxrt1189_cm33.overlay:
```diff
@@ -0,0 +1,9 @@
+ */
+
+temp_sensor: &temp {
+	status = "okay";
```
Set this in the board dts and remove the overlay

decsny commented:
@yperess @mmahadevan108 @DerekSnell @jeppenodgaard please review your blocking requests

jeppenodgaard commented on drivers/sensor/nxp/nxp_tmpsns/nxp_tmpsns.c:
```diff
@@ -0,0 +1,75 @@
+
+	TMPSNS_GetDefaultConfig(&config);
+	config.measureMode = kTEMPSENSOR_ContinuousMode;
+	config.frequency = 0x03U;
```
What does this mean? Consider adding a `#define`.

anedergaard replied:
Added a property in the binding for this driver, allowed for a bit more description to be added  than a define in my opinion.

jeppenodgaard reviewed (dismissed):
LGTM. Just a nit suggestion.

mmahadevan108 commented:
@yperess , can you revisit this PR.

MaureenHelm commented on dts/bindings/sensor/nxp,tmpsns.yaml:
```diff
@@ -0,0 +1,17 @@
+properties:
+  clock-divider:
+    type: int
+    default: 3
```
Need to explain in the description why the default value was selected
https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#dt-bindings-default-rules

anedergaard commented:
- Updated description of default values
- Rebased to main
- Ran TESTSUITE temp_sensor succeeded
<details><summary>Test Summary</summary>
<p>
*** Booting Zephyr OS build v4.2.0-1058-g591902d962f6 ***
Running TESTSUITE temp_sensor
===================================================================
START - test_polling
Temperature: 29.94
Temperature: 31.27
Temperature: 31.27
Temperature: 31.84
Temperature: 31.84
 PASS - test_polling in 2.010 seconds
===================================================================
START - test_trigger
This sensor does not support threshold trigger.
 SKIP - test_trigger in 0.005 seconds
===================================================================
TESTSUITE temp_sensor succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [temp_sensor]: pass = 1, fail = 0, skip = 1, total = 2 duration = 2.015 seconds
 - PASS - [temp_sensor.test_polling] duration = 2.010 seconds
 - SKIP - [temp_sensor.test_trigger] duration = 0.005 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
</p>
</details>

MaureenHelm reviewed (commented):
Please take a look at the twister failures

ZhaoxiangJin commented on modules/hal_nxp/mcux/mcux-sdk-ng/drivers/drivers.cmake:
```diff
@@ -143,6 +143,7 @@ set_variable_ifdef(CONFIG_MIPI_DBI_NXP_DCNANO_LCDIF CONFIG_MCUX_COMPONENT_driver
 set_variable_ifdef(CONFIG_MIPI_DBI_NXP_FLEXIO_LCDIF CONFIG_MCUX_COMPONENT_driver.flexio_mculcd)
 set_variable_ifdef(CONFIG_VIDEO_MCUX_MIPI_CSI2RX    CONFIG_MCUX_COMPONENT_driver.mipi_csi2rx)
 set_variable_ifdef(CONFIG_ETH_NXP_IMX_NETC          CONFIG_MCUX_COMPONENT_driver.netc)
+set_variable_ifdef(CONFIG_NXP_TMPSNS          CONFIG_MCUX_COMPONENT_driver.tempsensor)
```
Control indentation to be consistent with the previous line

ZhaoxiangJin commented on dts/arm/nxp/nxp_rt118x.dtsi:
```diff
@@ -1404,6 +1404,13 @@
 		status = "disabled";
 	};
+
+	temp: temp@4484580 {
```
why is it named temp here instead of tmpsns?

ZhaoxiangJin commented on dts/arm/nxp/nxp_rt118x.dtsi:
For devices that do not require a specific board setup, we usually define aliases in the SOC DT.
For example: https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/nxp/nxp_rt10xx.dtsi#L18

ZhaoxiangJin commented on dts/bindings/sensor/nxp,tmpsns.yaml:
```diff
@@ -0,0 +1,24 @@
+include: sensor-device.yaml
+
+properties:
+  clock-divider:
```
why is it named clock-divider here? DT properties should be named more specifically as for what the hardware is.
From RM, CTRL1[FREQ] represent sampling frequency.

ZhaoxiangJin reviewed (changes requested):
Hello @anedergaard, thanks for you contribution, i did not noticed this PR before, and i created the same [PR](https://github.com/zephyrproject-rtos/zephyr/pull/95039) 2 days before. Overall, this PR is LGTM, so i will close my PR and focus on this PR. I have add my comments, please clarify. Thanks.

ZhaoxiangJin reviewed (dismissed):
LGTM

ZhaoxiangJin commented on dts/bindings/sensor/nxp,tmpsns.yaml:
```diff
@@ -0,0 +1,17 @@
+properties:
+  freq:
+    type: int
+    default: 1000
```
This is not PoR reset value, @anedergaard, please add justifying here.

anedergaard replied:
I agree that the driver should not set a default value other then PoR reset value, I will rework and address the comments one of the first days in the coming work week.

anedergaard replied:
I've been testing using the `tests/drivers/sensor/temp_sensor` in polling mode, I have therefore decided to remove the automatic resampling of the tempsns since the driver does not support trigger in the current state.

If adding the trigger support in a later PR, it will probably also fit better to be set via the `sensor_attr_set` API.

anedergaard commented:
Updated to poll based only, testing using `tests/drivers/sensor/temp_sensor`
In `channel_fetch` the driver stops and starts the measuring sequence in order to ensure a new sample is measured and not one from a previous request.

Test output:
```shell
*** Booting Zephyr OS build v4.2.0-3013-g34287963c161 ***
Running TESTSUITE temp_sensor
===================================================================
START - test_polling
Temperature: 31.46
Temperature: 31.46
Temperature: 31.46
Temperature: 31.65
Temperature: 31.65
 PASS - test_polling in 2.010 seconds
===================================================================
START - test_trigger
This sensor does not support threshold trigger.
 SKIP - test_trigger in 0.005 seconds
===================================================================
TESTSUITE temp_sensor succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [temp_sensor]: pass = 1, fail = 0, skip = 1, total = 2 duration = 2.015 seconds
 - PASS - [temp_sensor.test_polling] duration = 2.010 seconds
 - SKIP - [temp_sensor.test_trigger] duration = 0.005 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```
